### DISCUSSION TITLE
Resetting the Com Port in bonsai

### DISCUSSION THREAD

**Discussion by benhardin13** [bonsai-rx/bonsai#869] [Q&A]
Hey everyone,

I am currently working on a project where I am collecting data from different animals. To do this I have a code on an Arduino running in the background that bonsai collects from. Whenever I change animals I want the port to reset. Is there a way to do this within bonsai? Or do I have to open and close my serial monitor on the Arduino to reset the com port?

**Comment by bruno-f-cruz** [comment]
Heya.

Would one of these work?

#### 1. Code a reset callback function
`void(* resetFunc) (void) = 0;`
You can then call `resetFunc` following the reception of a given serial instruction. You could use Bonsai's `SerialStringWrite` to achieve this.

#### 2. Enable DTR
Arduino resets when it receives a [DTR instruction](https://rheingoldheavy.com/arduino-from-scratch-part-11-atmega328p-dtr-and-reset/). Bonsai implements such an instruction when setting up a new serial device, and you can leverage this to reset an arduino each time you `CreateSerialPort`. Just make sure you set the `DtrEnable` property to `True`. In order to close and open a new serial communication with the same Arduino you can use `CreateObservable` followed by a `Switch`. The intuition of this last node is something like: "Each time a new observable is created I will close the previous one". This prevents errors that might result from trying to open a port that is already opened. Check the attached workflow for a very barebones implementation.

Hope it works,
Bruno
[Arduino_DTR.zip](https://github.com/bonsai-rx/bonsai/files/8446306/Arduino_DTR.zip)
